staging/src/k8s.io/kube-aggregator/pkg/controllers/status/available_controller.go

```diff
@@ -449,33 +458,57 @@ func (c *AvailableConditionController) deleteAPIService(obj interface{}) {
 		}
 	}
 	klog.V(4).Infof("Deleting %q", castObj.Name)
-	c.enqueue(castObj)
+	c.updateCacheForAPIService(castObj)
+	c.enqueueService(castObj)
 }
 
-// there aren't very many apiservices, just check them all.
-func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []*apiregistration.APIService {
+func (c *AvailableConditionController) getAPIServicesFor(obj runtime.Object) []string {
 	metadata, err := meta.Accessor(obj)
 	if err != nil {
 		utilruntime.HandleError(err)
 		return nil
 	}
+	c.RLock()
+	defer c.RUnlock()
+	if svcMap, ok := c.cache[metadata.GetNamespace()]; ok {
+		if names, exist := svcMap[metadata.GetName()]; exist {
+			return names
+		}
+	}
+	return nil
+}
 
-	var ret []*apiregistration.APIService
+func (c *AvailableConditionController) updateCacheForAPIService(obj runtime.Object) []string {
```

Comment:
simplify to `rebuildAPIServiceCache()`, list all APIServices and rebuild the entire cache, no argument or return value required